cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,8 +1309,168 @@ struct Inequality {
+        value_set = value_set->Slice(pivot);
+        break;
+      case Comparison::NOT_EQUAL:
+      case Comparison::NA:
+        DCHECK(false);
+        return Status::Invalid("Invalid comparison");
+    }
+
+    if (value_set->length() == 0) return false;
+    return value_set;
+  }
+
+  /// Simplify an `is_in` call against an inequality guarantee.
+  /// \pre `is_in_call` is a call to the `is_in` function
+  /// \post updates the simplification context with the simplified value set if
+  ///       nullopt is returned, otherwise ensures that the call is removed from
+  ///       the simplification context
+  /// \return a boolean if the whole `is_in` call simplifies to a boolean literal,
+  ///         otherwise nullopt
+  static Result<std::optional<bool>> SimplifyIsIn(
+      const Inequality& guarantee,
+      const Expression::Call* is_in_call,
+      SimplificationContext& context) {
+    DCHECK_EQ(is_in_call->function_name, "is_in");
+
+    // Null-matching behavior is complex and reduces the chances of reduction
+    // of `is_in` calls to a single literal for every possible input, so we
+    // abort the simplification if nulls are possible in the input.
+    if (guarantee.nullable) return std::nullopt;
+
+    if (!guarantee.bound.is_scalar()) {
+      return Status::Invalid("Cannot simplify inequality on a non-scalar bound");
+    }
+
+    const auto& lhs = Comparison::StripOrderPreservingCasts(is_in_call->arguments[0]);
+    if (!lhs.field_ref()) return std::nullopt;
+    if (*lhs.field_ref() != guarantee.target) return std::nullopt;
+
+    auto options = checked_pointer_cast<SetLookupOptions>(is_in_call->options);
+    std::array<TypeHolder, 2> types{guarantee.bound.type().get(),
+                                    options->value_set.type().get()};
+    TypeHolder cmp_type;
+    if (types[0] == types[1]) cmp_type = types[0];
+    if (!cmp_type) cmp_type = internal::CommonNumeric(types.data(), types.size());
+    if (!cmp_type) cmp_type = internal::CommonTemporal(types.data(), types.size());
+    if (!cmp_type) cmp_type = internal::CommonBinary(types.data(), types.size());
+    if (!cmp_type) return std::nullopt;
+
+    std::variant<std::shared_ptr<Array>, bool> result;
+
+#define CASE(TYPE_CLASS)                                                    \
+  case TYPE_CLASS##Type::type_id: {                                         \
+    ARROW_ASSIGN_OR_RAISE(                                                  \
+        std::shared_ptr<Array> value_set,                                   \
+        GetIsInValueSetForSimplification(is_in_call, cmp_type, context));   \
+    ARROW_ASSIGN_OR_RAISE(                                                  \
+        result,                                                             \
+        SimplifyIsInValueSet<TYPE_CLASS##Type>(guarantee, value_set));      \
+    break;                                                                  \
+  }
+
+    switch (cmp_type.id()) {
+      CASE(UInt8)
```

Comment:
Just a naive question, would decimal could be checked here? ( or just not support it yet?)

Reply:
It seems like supporting decimal types would be easy - the basic decimal C types support comparison operators and the `is_in` function has kernels for decimal types. 

But [here](https://github.com/apache/arrow/pull/43256#discussion_r1693913453) you mention that we should leave out floating point types, so should we support decimals but not floats? Or just support both?